File: iosApp/iosApp/Modules/MnemonicNew ✅/MnemonicNewWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultMnemonicNewWireframe {
    private weak var parent: UIViewController!
    private let context: MnemonicNewWireframeContext
    private let keyStoreService: KeyStoreService
    private let passwordService: PasswordService
    private let settingsService: SettingsService

    private weak var vc: UIViewController!

    init(
        _ parent: UIViewController?,
        context: MnemonicNewWireframeContext,
        keyStoreService: KeyStoreService,
        passwordService: PasswordService,
        settingsService: SettingsService
    ) {
        self.parent = parent
        self.context = context
        self.keyStoreService = keyStoreService
        self.passwordService = passwordService
        self.settingsService = settingsService
    }
}

// MARK: - MnemonicWireframe

extension DefaultMnemonicNewWireframe {

    func present() {
        let vc = wireUp()
        let presentingTopVc = (parent as? UINavigationController)?.topVc
        if settingsService.isSelected(setting: .init(group: .developer, action: .developerTransitionsCardFlip)) {
            let presentedTopVc = (vc as? UINavigationController)?.topVc
            let delegate = presentedTopVc as? UIViewControllerTransitioningDelegate
            self.vc = vc
            vc.modalPresentationStyle = .overFullScreen
            vc.transitioningDelegate = delegate
        } else if settingsService.isSelected(setting: .init(group: .developer, action: .developerTransitionsSheet)) {
            vc.modalPresentationStyle = .automatic
        }
        presentingTopVc?.present(vc, animated: true)
    }

    func navigate(with destination: MnemonicNewWireframeDestination) {
        if destination is MnemonicNewWireframeDestination.LearnMoreSalt {
            UIApplication.shared.open(Constant.saltExplanationURL)
        }
        if destination is MnemonicNewWireframeDestination.Dismiss {
            vc?.popOrDismiss()
        }
    }
}

private extension DefaultMnemonicNewWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultMnemonicNewInteractor(
            keyStoreService: keyStoreService,
            passwordService: passwordService
        )
        let vc: MnemonicNewViewController = UIStoryboard(.mnemonicNew).instantiate()
        let presenter = DefaultMnemonicNewPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.presenter = presenter
        let nc = NavigationController(rootViewController: vc)
        self.vc = nc
        return nc
    }
}

// MARK: - Constant

extension DefaultMnemonicNewWireframe {
    enum Constant {
        static let saltExplanationURL = URL(
            string: "https://www.youtube.com/watch?v=XqB5xA62gLw"
        )!
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/mnemonicnew/MnemonicNewWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicnew

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3walletcore.app.App
import com.sonsofcrypto.web3walletcore.modules.mnemonicNew.*
import com.sonsofcrypto.web3walletcore.services.password.PasswordService
import com.sonsofcrypto.web3walletcore.services.settings.SettingsService

class DefaultMnemonicNewWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: MnemonicNewWireframeContext,
    private val keyStoreService: KeyStoreService,
    private val passwordService: PasswordService,
): MnemonicNewWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.present(fragment, animated = true)
    }

    override fun navigate(destination: MnemonicNewWireframeDestination) {
        when (destination) {
            is MnemonicNewWireframeDestination.Dismiss -> {
                parent?.navigationFragment?.dismiss()
            }
            is MnemonicNewWireframeDestination.LearnMoreSalt -> {
                App.openUrl("https://www.youtube.com/watch?v=XqB5xA62gLw")
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = MnemonicNewFragment()
        val interactor = DefaultMnemonicNewInteractor(
            keyStoreService,
            passwordService,
        )
        val presenter = DefaultMnemonicNewPresenter(
            WeakRef(view),
            this,
            interactor,
            context,
        )
        view.presenter = presenter
        return view
    }
}